### PR TITLE
Simplify migrations

### LINKED ISSUE #326: Simplify migrations
The current migrations workflow is quite difficult to maintain.

The goal was to give full flexibility, but after some usage, it seems to make things more difficult.

Instead of having a migrations definition such as this:

```yaml

_migrations:
  - version: v1.0.0
    before:
      - &mig
        - "[[ _copier_conf.src_path / 'migrations.py' ]]"
        - "[[ _copier_conf|tojson ]]"
    after:
      - *mig
  - version: v2
    before: [*mig]
    after:
      - *mig
      - "rm delete-in-migration-$VERSION_CURRENT.txt"
```

We should be having something like this:

```yaml
_migrations:
  working_directory: "[[ _copier_conf.src_path ]]" # default value
  command: [poetry, run, migrations.py]
```

Then, all data needed for migrations will be passed as env variables, which that script can load and use to handle all of them comfortably.

At the end of the day, that code is easier to maintain in a migrations script, and this way we avoid the need to repeat the same definition once and again for each version (as seen in the 1st example using `*mig`).

### PR DESCRIPTION
**Pull request by AdrianFreundQC**
Resolves #326 

Here is a suggestion for how to simplify migrations. Some changes made to migrations were also applied to tasks for consistency.
The task configuration is 100% backwards compatible. The migration configuration is not backwards compatible. The old style should be seen as deprecated. This PR support both old and new style and shows a warning when the old style is used.

## Tasks

Tasks can still be configured like before:
```yaml
_tasks:
  - git init
```
  
  Additionally Tasks can be a dictionary, which allows passing one or more of the optional arguments `when` and `working_directory`. `working_directory` defaults to the destination directory and can be either relative to it, or absolute. Both arguments can be templated.
  
```yaml
_tasks:
  - command: git init
    working_directory: "submodule"
    when: "{{ enable_submodules }}"
```

## Migrations

It is no longer necessary to duplicate the same command line for every version to run a migration script.
```yaml
_migrations:
  - "{{ _copier_python }} {{ _copier_conf.src_path }}/migrate.py"
```
This runs twice (before and after) on every update and still passes environment variables to the script.
Note that it can't pass `$VERSION_(PEP440_)CURRENT` because there is no _current_ version. Only a from and to version.

It is still possible to only run migrations on a specific version:
```yaml
_migrations:
  - version: v1.4
    command: rm old_file.txt || true
```
In this case `$VERSION_(PEP440_)CURRENT` is still passed to the command. 

This still runs both before and after.
`working_directory` and `when` are also supported for migrations. In addition all the environment variables passed to the command are also available for use in templating:

```yaml
_migrations:
  - version: v1.4
    command: rm old_file.txt
    when: "{{ _stage == 'before' }}"
```

## Current State

All the features described here work.
I still want to do a bit of code cleanup and improve the documentation in the next few days.
I would also like some feedback on this design. Is there anything you don't agree with? Something I should change?

### CODE REVIEW

**Review by sisp — CHANGES_REQUESTED** [review]
Thanks for taking the initiative and working on this long-standing feature proposal, @AdrianFreundQC! :pray:

I've done a first review iteration over the main code and docs (not the tests yet). It looks great, just a couple of minor comments so far.

File: copier/template.py
Comment (AdrianFreundQC) [reply copier/template.py:390]: The one in the new format branch is only executed if a `version` is specified.
However, this is part of the code I wanted to clean up anyways.

File: docs/configuring.md
Comment (AdrianFreundQC) [reply docs/configuring.md:1159]: Yes. I added the jinja variables after writing the documentation. Will document them later.

File: docs/configuring.md
Comment (AdrianFreundQC) [reply docs/configuring.md:1199]: I know that it's different, but it's just an arbitrary example.
I just wanted to go with something that clearly shows how it works.

I guess I can add $STAGE, but in reality none of those would have to be passed as arguments, as they are available as env variables to the task.

**Comment by AdrianFreundQC** [comment]
I pushed the fixes as new commits to simplify review.
I can rebase them back into three commits (feat, test, docs) or you can just squash merge if you prefer that.

**Comment by AdrianFreundQC** [comment]
CI failure looks unrelated. Should be fixed with a re-run

File: docs/configuring.md
Comment (AdrianFreundQC) [inline docs/configuring.md:1210]: Or v9.2.0?

**Comment by freundTech** [comment]
One more thought I had: the migration command running twice on every update might be confusing and unexpected for new users, however it is required for more complex migrations.

Changing the default value of `when` from `true` to `{{ _stage == "after" }}` might solve this issue. Migrations without `when` only run once, while experienced users can still set `when` to `true` to have them run in the before and after stage.

**Comment by AdrianFreundQC** [comment]
One more thought I had: the migration command running twice on every update might be confusing and unexpected for new users, however it is required for more complex migrations.

Changing the default value for `when` from `true` to `{{ _state == "after" }}` might solve this issue. Migrations without `when` only run once, while experience users can still set `when` to `true` to have them run in the before and after stage.

**Comment by sisp** [comment]
I like this idea, it sounds like a better default behavior than running the command twice. 👌

By the way, I haven't forgotten about reviewing this PR, just haven't had enough time yet. 🙁 I'll try to do it on Monday. 🤞

File: tests/test_migrations.py
Comment (sisp) [inline tests/test_migrations.py:1]: I wonder whether we should rename the original version of this file to `test_migrations_legacy.py` and create a copy named `test_migrations.py` which contains the same tests using the new format _plus_ additional tests covering new aspects like:

* The availability of Jinja variables equivalent to the environment variables.
* The default `when` behavior (e.g., `when: "{{ _stage == 'after' }}"` [as you suggested](https://github.com/copier-org/copier/pull/1510#issuecomment-1962313028).
* The behavior of, e.g., `when: true` and possibly other typical cases.
* The default working directory.
* The behavior of a custom working directory.
* More?

This way, we can ensure:

* The legacy format remains functional as tested by the original tests.
* The original tests adapted to the new format also pass for the new format.
* The additional capabilities of the new format are tested.
* When we decide to drop support for the legacy format, we only need to delete `test_migrations_legacy.py` and the demo fixture instead of surgically removing code associated with the legacy format.

WDYT, @AdrianFreundQC and @copier-org/maintainers?

File: tests/test_tasks.py
Comment (sisp) [inline tests/test_tasks.py:1]: Same thought as for `test_migrations.py`.

**Review by sisp — CHANGES_REQUESTED** [review]
I've done a full review this time. The implementation looks good to me. The legacy format test does not ensure identical behavior as before anymore, and I think the new format needs a bit more systematic testing. I think keeping the legacy tests unchanged and adapting and extending a copy thereof for the new format would be better than surgically adding some testing of the new format and mixing it with the old format (see my inline comment for more details).

File: docs/configuring.md
Comment (yajo) [inline docs/configuring.md:1158]: suggestion(docs): I think it's worth to mention that using ˋversionˋ can still save you from being marked as an unsafe template for every update, because the unsafe part will only happen in *that* specific update, and not in *any* update.

**Review by yajo — CHANGES_REQUESTED** [review]
This definitely looks like the path forward.  Thanks for taking care of all this!

File: docs/configuring.md
Comment (sisp) [reply docs/configuring.md:1180]: Or simply `_task` dict? "env" carries some semantics of environment variables IMO.

File: copier/main.py
Comment (AdrianFreundQC) [reply copier/main.py:288]: I don't think that's needed, but it could make the code a bit clearer.
The current solution also allows using absolute paths:
```python
>>> from pathlib import Path
>>> p = Path("/user/test")
>>> p
PosixPath('/user/test')
>>> p / "/user/other"
PosixPath('/user/other')
```

File: docs/configuring.md
Comment (AdrianFreundQC) [reply docs/configuring.md:1223]: They have once this is merged, but this part talks about older versions.
The only case where you can use the Jinja variables is if you already have the new version, but are using the old configuration format.

File: docs/configuring.md
Comment (pawamoy) [reply docs/configuring.md:1158]: Also "if not version" -> "if no version".

File: docs/configuring.md
Comment (yajo) [reply docs/configuring.md:1210]: Yes

File: copier/main.py
Comment (yajo) [reply copier/main.py:288]: Ah I didn't know that, thanks.

File: docs/configuring.md
Comment (yajo) [reply docs/configuring.md:1180]: Well, both options have pros and cons, so I think it's fair to respect your strong preference, since you're the one contributing this :)

File: docs/configuring.md
Comment (hparfr) [inline docs/configuring.md:1428]: thanks for this work

>Commands to execute after generating or updating a project from your template.

Because tasks are run after generating AND updating, is it possible to test this in the "when" clause ?

```yaml
- command:  git clone something in_directory/
  when : `copier copy`

- command: cd in_directory && git pull
  when:  `copier update`
```

?

Comment (sisp) [reply docs/configuring.md:1428]: Work on supporting pre/post copy/update tasks has started in #1511 and will probably build upon this PR.

File: docs/configuring.md
Comment (AdrianFreundQC) [reply docs/configuring.md:1451]: I just copied the command from the old version, where it already used a `/`.
I don't know if PowerShell can also deal with `/`, but I can just change it to `\`

**Comment by AdrianFreundQC** [comment]
Sorry for the delay. I've been quite busy recently.
I'm still working on the tests for the new migrations implementation.
I'll have some time in the next two weeks and if that's not enough I'll probably be able to finish it in the first half of May.

**Comment by sisp** [comment]
Thanks for the update and your continuous work on this PR, @AdrianFreundQC! :pray:

**Comment by AdrianFreundQC** [comment]
I have no idea why the MacOS flake test is suddenly failing now. Maybe a broken cache?

> /Users/runner/work/_temp/be5ceacc-d596-4a1f-b513-54880e12e09d.sh: /Users/runner/work/copier/copier/.venv/bin/copier: /Users/runner/work/copier/copier/.venv/bin/python: bad interpreter: No such file or directory

The flake check runs fine on my machine.

**Comment by sisp** [comment]
@AdrianFreundQC Sorry for the inconvenience. It's because `macos-latest` on GitHub-hosted runners got bumped to macOS 14 (ARM) for which there are no binaries for Python 3.8 and 3.9. Before, `macos-latest` was referring to macOS 12 (x64) for which those binaries existed. See https://github.com/copier-org/copier/pull/1602#pullrequestreview-2019446447 for more details.

We're waiting for @yajo to resolve #1604 (because of a project-level settings change only he can make). Then, after rebasing your PR, CI checks should work again.

File: docs/configuring.md
Comment (AdrianFreundQC) [reply docs/configuring.md:1158]: Done.

**Comment by AdrianFreundQC** [comment]
I think all discussions should have been addressed now and the CI seems to be working (assuming my last change didn't break anything).

File: tests/test_migrations.py
Comment (AdrianFreundQC) [reply tests/test_migrations.py:1]: I kept the legacy test as is, but started from scratch for the new tests, to produce more maintainable and better structured tests. I think they should cover everything.

**Comment by AdrianFreundQC** [comment]
I think all discussions should have been addressed now and the CI seems to be working (assuming my last change didn't break anything).

File: docs/configuring.md
Comment (yajo) [inline docs/configuring.md:1193]: suggestion: it'd be cool to link to docs of this object if possible.

```suggestion
    represented as [packaging.version.Version](https://packaging.pypa.io/en/stable/version.html#packaging.version.Version) objects and allow access to their
```

**Review by yajo — APPROVED** [review]
Very amazing work, Thanks!

There's just one minor suggestion to enhance docs. Would you like to push it before merging?

File: docs/configuring.md
Comment (AdrianFreundQC) [reply docs/configuring.md]: Done!

**Review by sisp — CHANGES_REQUESTED** [review]
Looks great, @AdrianFreundQC! :ok_hand: Just some totally minor remarks from my side. Plus in general, some lines seem very long. Would you mind running Ruff's formatter once?

About code formatting: IIRC, there is no Nix-based pre-commit hook for Ruff's formatter, which would explain why the `flake-check` checks pass although I'm sure Ruff's formatter would disagree. @yajo Is there a way to enable Ruff's formatter to check code formatting? And does the Nix-locked version of Ruff even have the formatter already? We might need to update the lock file.

**Comment by AdrianFreundQC** [comment]
I ran addressed the two comments above and ran `ruff format` (nix pinned version) on `main.py` and `template.py`. I didn't run it on all files to not clutter this PR.

I also rebased on master again.

File: tests/test_migrations.py
Comment (sisp) [inline tests/test_migrations.py:285]: One more empty line here:

```suggestion


```

File: tests/test_migrations.py
Comment (sisp) [inline tests/test_migrations.py:317]: One more empty line here:

```suggestion


```

File: tests/test_migrations.py
Comment (sisp) [inline tests/test_migrations.py:402]: One more empty line here:

```suggestion


```

File: tests/test_migrations.py
Comment (sisp) [inline tests/test_migrations.py:404]: This line is too long. Could you run Ruff's formatter on this file?

File: tests/test_migrations.py
Comment (sisp) [inline tests/test_migrations.py:458]: This line is too long. Could you run Ruff's formatter on this file?

File: tests/test_tasks.py
Comment (sisp) [inline tests/test_tasks.py:47]: This line is too long. Could you run Ruff's formatter on this file?

File: tests/test_migrations.py
Comment (sisp) [inline tests/test_migrations.py:475]: This line is too long. Could you run Ruff's formatter on this file?

**Review by sisp — CHANGES_REQUESTED** [review]
Sorry for the delay. 🫣

Content-wise it looks good to me. I still see some minor formatting issues. Would you mind trying to fix them?

**Comment by AdrianFreundQC** [comment]
No problem.
I reformatted the files and rebased on master again.

**Review by sisp — APPROVED** [review]
Looks good now! :tada: Thank you very much for the hard work, implementing this great feature with proper tests, and patience with the reviews even on pedantic formatting fixes! :pray: Fantastic job! :1st_place_medal: Let's get this merged! :rocket:

**Comment by AdrianFreundQC** [comment]
Do you have an ETA on when the next version might get released?

**Comment by yajo** [comment]
I'd like to at least have https://github.com/copier-org/copier/issues/1598 fixed before release. It appeared because of this merge IIRC. It should be quite easy to fix though, I just had very little time lately. Once that one's fixed I'll push the new release. Would you want to contribute it?

**Comment by yajo** [comment]
Ah, nevermind. I just opened https://github.com/copier-org/copier/pull/1681. It was an easy one.

**Pull request merged**
